### PR TITLE
📦 [Dependencies.swift] Add `automaticProductType` to `SwiftPackageManagerDependencies`

### PR DESCRIPTION
**Pull request by danyf90**
### Short description 📝

Solves this comment: https://github.com/tuist/tuist/pull/3099#discussion_r657672195

### Checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase.
- [ ] The changes have been tested following the [documented guidelines](https://docs.tuist.io/contributors/testing-strategy/).
- [ ] The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ] In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

File: Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift
Comment (fortmarek) [inline Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift:14]: I'm not sure if this should be a "global" switch - I could imagine people would want to configure this on a package-by-package basis. Plus if we did it this way, you wouldn't be able to do what SPM automatic linking does (that is link dynamically only libraries that are necessary).

Comment (danyf90) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift:14]: Makes sense, I'll add it to the `Package` then 🙌

Comment (danyf90) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift:14]: the problem is that we will have also to handle packages and targett not defined here (e.g. dependencies).
I'll replace the `automaticProductType: Product` here with a `customProductType: [String: Product]` which will allow to map any target to a custom product if needed.

Makes sense?

Comment (fortmarek) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift:14]: ahh, right, I guess that's the most bulletproof approach. cc @ollieatkinson @laxmorek for possible additional thoughts (especially @ollieatkinson has already done some research regarding the automatic linking)

Comment (danyf90) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift]: Changed here: https://github.com/tuist/tuist/pull/3105/commits/6eeccc63e6e409cbe3a082174fdb5e891ad7fad9

File: Sources/TuistDependencies/SwiftPackageManager/Utils/SwiftPackageManagerGraphGenerator.swift
Comment (fortmarek) [inline Sources/TuistDependencies/SwiftPackageManager/Utils/SwiftPackageManagerGraphGenerator.swift:217]: nit: unnecessary line

**Review by fortmarek — APPROVED** [review]
This is good from my side, but I'd wait for @ollieatkinson's validation before merging

**Comment by ollieatkinson** [comment]
Would you be able to add a test case showing how you envisage the product types being defined? All of the test cases here are passing an empty dictionary

File: Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift
Comment (ollieatkinson) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift]: Being able to automatically determine the linking type is something Tuist should offer through Dependencies.swift, however, I like what is done here as a way to specify an override behaviour when the automatic choice does not meet your requirement. If this is purely additive and not required to specify for each package dependency then the change looks good 👍 Thanks @danyf90

**Comment by danyf90** [comment]
> Would you be able to add a test case showing how you envisage the product types being defined? All of the test cases here are passing an empty dictionary

Added one scenario of overriding the default to `SwiftPackageManagerGraphGeneratorTests` here: https://github.com/tuist/tuist/pull/3105/commits/7c521805b6ba7f47c4a8dac8a09fb0df7d718b59

File: Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift
Comment (danyf90) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift]: Ideally yes, but we would need a way here (i.e. in `ProjectDescription.TargetDependency` and `TuistGraph.TargetDependency`) to specify something like `automaticFramework` and then should be solved later on during graph generation, as the information is not available yet at this point of the pipeline.

I think that defaulting to staticFramework and allowing to override it in the `Dependencies.swift` should be enough for now. I have opened [an issue](https://github.com/tuist/tuist/projects/6#card-64009985) to keep track of the limitation

Comment (fortmarek) [reply Sources/ProjectDescription/Dependencies/SwiftPackageManagerDependencies.swift]: sounds good @danyf90 👍

**Comment by fortmarek** [comment]
I think we can merge this once CI is green

**Pull request merged**
